runtime/compiler/env/J9ClassEnv.cpp

```diff
@@ -825,6 +825,29 @@ J9::ClassEnv::isValueTypeClassFlattened(TR_OpaqueClassBlock *clazz)
    return (clazz && J9_IS_J9CLASS_FLATTENED(reinterpret_cast<J9Class*>(clazz)));
    }
 
+bool
+J9::ClassEnv::isValueBasedClass(TR_OpaqueClassBlock *clazz)
+   {
+#if defined(J9VM_OPT_JITSERVER)
+   if (auto stream = TR::CompilationInfo::getStream())
+      {
+      uintptr_t classFlags = 0;
+      JITServerHelpers::getAndCacheRAMClassInfo((J9Class *)clazz, TR::compInfoPT->getClientData(), stream, JITServerHelpers::CLASSINFO_CLASS_FLAGS, (void *)&classFlags);
+#ifdef DEBUG
+      stream->write(JITServer::MessageType::ClassEnv_classFlagsValue, clazz);
+      uintptr_t classFlagsRemote = std::get<0>(stream->read<uintptr_t>());
+      // Check that class flags from remote call is equal to the cached ones
+      classFlags = classFlags & J9ClassIsValueBased;
+      classFlagsRemote = classFlagsRemote & J9ClassIsValueBased;
+      TR_ASSERT(classFlags == classFlagsRemote, "remote call class flags is not equal to cached class flags");
```

Comment:
Should be a `TR_FATAL_ASSERT`, especially under debug.

Reply:
As far as I could recall it's possible that these asserts could be hit on the server side. I don't remember how frequent it was or if things have changed. The same kind of `TR_ASSERT` check is also used in other four methods of `ClassEnv` [4]. I'll let @mpirvu help comment on if it should be changed to `TR_FATAL_ASSERT` based on the latest JITServer development. 

[1]
https://github.com/eclipse/openj9/blob/f10260975a1d840fc8355bb70c4a1d42a9303337/runtime/compiler/env/J9ClassEnv.cpp#L125
https://github.com/eclipse/openj9/blob/f10260975a1d840fc8355bb70c4a1d42a9303337/runtime/compiler/env/J9ClassEnv.cpp#L783
https://github.com/eclipse/openj9/blob/f10260975a1d840fc8355bb70c4a1d42a9303337/runtime/compiler/env/J9ClassEnv.cpp#L806
https://github.com/eclipse/openj9/blob/f10260975a1d840fc8355bb70c4a1d42a9303337/runtime/compiler/env/J9ClassEnv.cpp#L829

Reply:
@a7ehuo Thank you for not forgetting about JITServer!
The check is there to validate that the flag the server has cached is still consistent with the client's view.
This needs to be done on a case-by-case basis, meaning that there are some flags which are expected to be set once and never change. The check is valid for such flags.
There are other flags that could change back and forth at runtime and we should not put a check for those.
The question here is whether `J9ClassIsValueBased` can change at runtime, or it is set during class load and immutable afterwards.
Irrespective of the answer, since the checking code is under `#ifdef DEBUG` TR_ASSERT has the same behavior as TR_ASSERT_FATAL, so it doesn't matter which one we use. I am ok with changing this into a fatal assert.

Reply:
Once `J9ClassIsValueBased` or `J9ClassIsValueType` is set, they're immutable afterwards. I'll change the assert to TR_ASSERT_FATAL